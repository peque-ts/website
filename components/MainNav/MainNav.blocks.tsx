import clsx from 'clsx';
import Link from 'next/link';
import React from 'react';

import { MAIN_NAV_ITEMS, useNavMobileButton } from './MainNav.helpers';

const MainNavItems: React.VFC = () => (
  <>
    {MAIN_NAV_ITEMS.map(({ name, to }, index) => (
      <li key={index}>
        <Link href={to}>
          <a className="text-lg">{name}</a>
        </Link>
      </li>
    ))}
  </>
);

const MainNavMobile: React.VFC<{ bgClassName?: string }> = (props) => {
  const { open, renderButton } = useNavMobileButton();

  return (
    <div className="block tablet:hidden" data-testid="MainNavMobile">
      {renderButton()}
      <div
        className={clsx(
          'absolute inset-x-0 -z-10 shadow-lg transition',
          props.bgClassName,
          open ? 'translate-y-0' : '-translate-y-full',
        )}
      >
        <ul className={clsx('flex-col items-center space-y-4 pt-4 pb-8', open ? 'flex' : 'hidden')}>
          <MainNavItems />
        </ul>
      </div>
    </div>
  );
};

const MainNavDesktop: React.VFC = () => (
  <ul className="hidden space-x-8 tablet:flex" data-testid="MainNavDesktop">
    <MainNavItems />
  </ul>
);

export { MainNavMobile, MainNavDesktop };
