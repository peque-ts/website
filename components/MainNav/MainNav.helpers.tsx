import clsx from 'clsx';
import React, { useCallback, useState } from 'react';

import { NavItem } from '../../lib/nav.types';

export const MAIN_NAV_ITEMS: Pick<NavItem, 'name' | 'to'>[] = [
  {
    name: 'Docs',
    to: '/docs',
  },
  {
    name: 'Enterprise',
    to: '/enterprise',
  },
  {
    name: 'Sponsors',
    to: '/sponsors',
  },
];

export const useNavMobileButton = () => {
  const [open, setOpen] = useState(false);

  const renderButton = useCallback(() => {
    const renderSegment = (className: string) => (
      <span
        aria-hidden="true"
        className={clsx(
          'absolute block h-0.5 w-5 transform bg-white transition ease-in-out',
          className,
        )}
      />
    );

    return (
      <button
        type="button"
        className="relative h-10 w-10 focus:outline-none"
        onClick={() => setOpen((value) => !value)}
      >
        <span className="sr-only">Open main menu</span>
        <div className="absolute left-1/2 top-1/2 block w-5 -translate-x-1/2 -translate-y-1/2 transform">
          {renderSegment(open ? 'rotate-45' : '-translate-y-1.5')}
          {renderSegment(open ? 'opacity-0' : '')}
          {renderSegment(open ? '-rotate-45' : 'translate-y-1.5')}
        </div>
      </button>
    );
  }, [open]);

  return { renderButton, open };
};
