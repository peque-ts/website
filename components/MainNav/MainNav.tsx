import React from 'react';

import { MainNavDesktop, MainNavMobile } from './MainNav.blocks';

interface Props {
  bgClassName?: string;
}

export const MainNav: React.VFC<Props> = ({ bgClassName }) => (
  <nav>
    <MainNavMobile bgClassName={bgClassName} />
    <MainNavDesktop />
  </nav>
);
