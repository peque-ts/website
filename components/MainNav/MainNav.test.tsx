import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { MainNav } from './MainNav';

describe('MainNav', () => {
  it('should render', () => {
    const { asFragment } = render(<MainNav />);

    expect(screen.getByTestId('MainNavDesktop')).toBeInTheDocument();
    expect(screen.getByTestId('MainNavMobile')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should toggle mobile navigation', () => {
    const { asFragment } = render(<MainNav />);

    expect(screen.getByTestId('MainNavMobile')).toBeInTheDocument();

    // assert hidden by default
    expect(within(screen.getByTestId('MainNavMobile')).getByRole('list')).toHaveClass('hidden');

    // expand menu
    userEvent.click(screen.getByRole('button'));

    // assert no longer hidden
    expect(within(screen.getByTestId('MainNavMobile')).getByRole('list')).not.toHaveClass('hidden');
    expect(asFragment()).toMatchSnapshot();

    // collapse menu
    userEvent.click(screen.getByRole('button'));

    // assert hidden again
    expect(within(screen.getByTestId('MainNavMobile')).getByRole('list')).toHaveClass('hidden');

    expect(asFragment()).toMatchSnapshot();
  });
});
