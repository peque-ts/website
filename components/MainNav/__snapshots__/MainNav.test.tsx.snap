// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainNav should render 1`] = `
<DocumentFragment>
  <nav>
    <div
      class="block tablet:hidden"
      data-testid="MainNavMobile"
    >
      <button
        class="relative h-10 w-10 focus:outline-none"
        type="button"
      >
        <span
          class="sr-only"
        >
          Open main menu
        </span>
        <div
          class="absolute left-1/2 top-1/2 block w-5 -translate-x-1/2 -translate-y-1/2 transform"
        >
          <span
            aria-hidden="true"
            class="absolute block h-0.5 w-5 transform bg-white transition ease-in-out -translate-y-1.5"
          />
          <span
            aria-hidden="true"
            class="absolute block h-0.5 w-5 transform bg-white transition ease-in-out"
          />
          <span
            aria-hidden="true"
            class="absolute block h-0.5 w-5 transform bg-white transition ease-in-out translate-y-1.5"
          />
        </div>
      </button>
      <div
        class="absolute inset-x-0 -z-10 shadow-lg transition -translate-y-full"
      >
        <ul
          class="flex-col items-center space-y-4 pt-4 pb-8 hidden"
        >
          <li>
            <a
              class="text-lg"
              href="/docs"
            >
              Docs
            </a>
          </li>
          <li>
            <a
              class="text-lg"
              href="/enterprise"
            >
              Enterprise
            </a>
          </li>
          <li>
            <a
              class="text-lg"
              href="/sponsors"
            >
              Sponsors
            </a>
          </li>
        </ul>
      </div>
    </div>
    <ul
      class="hidden space-x-8 tablet:flex"
      data-testid="MainNavDesktop"
    >
      <li>
        <a
          class="text-lg"
          href="/docs"
        >
          Docs
        </a>
      </li>
      <li>
        <a
          class="text-lg"
          href="/enterprise"
        >
          Enterprise
        </a>
      </li>
      <li>
        <a
          class="text-lg"
          href="/sponsors"
        >
          Sponsors
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;

exports[`MainNav should toggle mobile navigation 1`] = `
<DocumentFragment>
  <nav>
    <div
      class="block tablet:hidden"
      data-testid="MainNavMobile"
    >
      <button
        class="relative h-10 w-10 focus:outline-none"
        type="button"
      >
        <span
          class="sr-only"
        >
          Open main menu
        </span>
        <div
          class="absolute left-1/2 top-1/2 block w-5 -translate-x-1/2 -translate-y-1/2 transform"
        >
          <span
            aria-hidden="true"
            class="absolute block h-0.5 w-5 transform bg-white transition ease-in-out rotate-45"
          />
          <span
            aria-hidden="true"
            class="absolute block h-0.5 w-5 transform bg-white transition ease-in-out opacity-0"
          />
          <span
            aria-hidden="true"
            class="absolute block h-0.5 w-5 transform bg-white transition ease-in-out -rotate-45"
          />
        </div>
      </button>
      <div
        class="absolute inset-x-0 -z-10 shadow-lg transition translate-y-0"
      >
        <ul
          class="flex-col items-center space-y-4 pt-4 pb-8 flex"
        >
          <li>
            <a
              class="text-lg"
              href="/docs"
            >
              Docs
            </a>
          </li>
          <li>
            <a
              class="text-lg"
              href="/enterprise"
            >
              Enterprise
            </a>
          </li>
          <li>
            <a
              class="text-lg"
              href="/sponsors"
            >
              Sponsors
            </a>
          </li>
        </ul>
      </div>
    </div>
    <ul
      class="hidden space-x-8 tablet:flex"
      data-testid="MainNavDesktop"
    >
      <li>
        <a
          class="text-lg"
          href="/docs"
        >
          Docs
        </a>
      </li>
      <li>
        <a
          class="text-lg"
          href="/enterprise"
        >
          Enterprise
        </a>
      </li>
      <li>
        <a
          class="text-lg"
          href="/sponsors"
        >
          Sponsors
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;

exports[`MainNav should toggle mobile navigation 2`] = `
<DocumentFragment>
  <nav>
    <div
      class="block tablet:hidden"
      data-testid="MainNavMobile"
    >
      <button
        class="relative h-10 w-10 focus:outline-none"
        type="button"
      >
        <span
          class="sr-only"
        >
          Open main menu
        </span>
        <div
          class="absolute left-1/2 top-1/2 block w-5 -translate-x-1/2 -translate-y-1/2 transform"
        >
          <span
            aria-hidden="true"
            class="absolute block h-0.5 w-5 transform bg-white transition ease-in-out -translate-y-1.5"
          />
          <span
            aria-hidden="true"
            class="absolute block h-0.5 w-5 transform bg-white transition ease-in-out"
          />
          <span
            aria-hidden="true"
            class="absolute block h-0.5 w-5 transform bg-white transition ease-in-out translate-y-1.5"
          />
        </div>
      </button>
      <div
        class="absolute inset-x-0 -z-10 shadow-lg transition -translate-y-full"
      >
        <ul
          class="flex-col items-center space-y-4 pt-4 pb-8 hidden"
        >
          <li>
            <a
              class="text-lg"
              href="/docs"
            >
              Docs
            </a>
          </li>
          <li>
            <a
              class="text-lg"
              href="/enterprise"
            >
              Enterprise
            </a>
          </li>
          <li>
            <a
              class="text-lg"
              href="/sponsors"
            >
              Sponsors
            </a>
          </li>
        </ul>
      </div>
    </div>
    <ul
      class="hidden space-x-8 tablet:flex"
      data-testid="MainNavDesktop"
    >
      <li>
        <a
          class="text-lg"
          href="/docs"
        >
          Docs
        </a>
      </li>
      <li>
        <a
          class="text-lg"
          href="/enterprise"
        >
          Enterprise
        </a>
      </li>
      <li>
        <a
          class="text-lg"
          href="/sponsors"
        >
          Sponsors
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
